cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs

```diff
@@ -0,0 +1,353 @@
+
+-- | The readonly environment
+data ROEnv txid tx = ROEnv
+  { targetBacklog     :: Int -- ^ how many to try to keep in back pocket, >0
+  , txNumServiceTime  :: Maybe DiffTime -- ^ seconds per tx
+  , txSizeServiceTime :: Maybe DiffTime -- ^ seconds per tx octet
+  } deriving Show
+
+defaultROEnv :: ROEnv txid tx
+defaultROEnv = ROEnv
+  { targetBacklog     = 1
+  , txNumServiceTime  = Nothing
+  , txSizeServiceTime = Nothing
+  }
+
+data RWEnv m txid tx = RWEnv
+  { terminating     :: Bool
+  , activityState   :: ActivityState
+  , proceedAfter    :: Time m
+  , availableOp     :: Maybe (Int, LazyTMVar m (Maybe [(txid, TxSubmit.TxSizeInBytes)]))
+  -- ^ the window and the response action
+  , inFlight
+  , notYetSent      :: Seq (txid, tx,  TxSubmit.TxSizeInBytes)
+  }
+
+defaultRWEnv :: MonadTime m => m (RWEnv m txid tx)
+defaultRWEnv = do
+  now <- getMonotonicTime
+  pure $ RWEnv
+    { terminating     = False
+    , activityState   = Idle
+    , proceedAfter    = now
+    , availableOp     = Nothing
+    , inFlight        = mempty
+    , notYetSent      = mempty
+    }
+
+-- | RPC interaction with `TxSubmit.TxSubmission`
+data RPCTxSubmission m txid tx
+  = RPCRequestTxIdsPromptly (Word16, Word16) (LazyTMVar m [(txid, TxSubmit.TxSizeInBytes)])
+  -- ^ Request contains the acknowledged number and the size of the
+  --   open window. Response contains the list of transactions (that
+  --   can be empty - see the `TxSubmit.TxSubmission` description of
+  --   `TxSubmit.StBlockingStyle` for more details). A prompt response
+  --   is expected.
+  |  RPCRequestTxIds (Word16, Word16) (LazyTMVar m (Maybe [(txid, TxSubmit.TxSizeInBytes)]))
+  -- ^ Request contains the acknowledged number and the size of the
+  --   open window. Response contains the list of transactions (that
+  --   can not be empty - see the `TxSubmit.TxSubmission` description
+  --   of `TxSubmit.StBlockingStyle` for more details); `Nothing`
+  --   indicates no more transaction submissions and a clean
+  --   shutdown. A prompt response is not expected.
+  | RPCRequestTxs [txid] (LazyTMVar m [tx])
+  -- ^ Request contains the list of transaction identifiers which are
+  --   returned in the response.
+
+-- | Tracer
+data TraceBenchTxSubmit txid
+  = TraceBenchTxSubRecv [txid]
+  -- ^ Received from generator.
+  | TraceBenchTxSubStart [txid]
+  -- ^ The @txid@ has been submitted to `TxSubmit.TxSubmission`
+  --   protocol peer.
+  | TraceBenchTxSubServReq [txid]
+  -- ^ Request for @tx@ recieved from `TxSubmit.TxSubmission` protocol
+  --   peer.
+  | TraceBenchTxSubServAck [txid]
+  -- ^ An ack (window moved over) received for these transactions.
+  | TraceBenchTxSubIdle
+  -- ^ Remote peer requested new transasctions but none were
+  --   available, generator not keeping up?
+  | TraceBenchTxSubRateLimit DiffTime
+  -- ^ Rate limiter bit, this much delay inserted to keep within
+  --   configured rate.
+  | TraceBenchTxSubDebug String
```

Comment:
Can we doc this constructor too?

Reply:
I'm not sure if we need this constructor at all. It was used as a fastest way to get raw debug messages..